staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation_test.go

```diff
@@ -658,3 +663,857 @@ func TestCelCompilation(t *testing.T) {
+						Format: passedFormat,
+					},
+					Extensions: schema.Extensions{
+						XIntOrString: xIntString,
+					},
+				}},
+			},
+			ValueValidation: &schema.ValueValidation{
+				MaxProperties: maxProperties,
+			},
+			Extensions: schema.Extensions{
+				XValidations: apiextensions.ValidationRules{
+					{
+						Rule: rule,
+					},
+				},
+			},
+		}
+	}
+}
+
+func genStringWithRule(rule string) func(maxLength *int64) *schema.Structural {
+	return func(maxLength *int64) *schema.Structural {
+		return &schema.Structural{
+			Generic: schema.Generic{
+				Type: "string",
+			},
+			ValueValidation: &schema.ValueValidation{
+				MaxLength: maxLength,
+			},
+			Extensions: schema.Extensions{
+				XValidations: apiextensions.ValidationRules{
+					{
+						Rule: rule,
+					},
+				},
+			},
+		}
+	}
+}
+
+func genBytesWithRule(rule string) func(maxLength *int64) *schema.Structural {
+	return func(maxLength *int64) *schema.Structural {
+		return &schema.Structural{
+			Generic: schema.Generic{
+				Type: "string",
+			},
+			ValueValidation: &schema.ValueValidation{
+				MaxLength: maxLength,
+				Format:    "byte",
+			},
+			Extensions: schema.Extensions{
+				XValidations: apiextensions.ValidationRules{
+					{
+						Rule: rule,
+					},
+				},
+			},
+		}
+	}
+}
+
+func genNestedSpecWithRule(rule string) func(maxLength *int64) *schema.Structural {
+	return func(maxLength *int64) *schema.Structural {
+		return &schema.Structural{
+			Generic: schema.Generic{
+				Type: "object",
+				AdditionalProperties: &schema.StructuralOrBool{Structural: &schema.Structural{
+					Generic: schema.Generic{
+						Type: "string",
+					},
+					ValueValidation: &schema.ValueValidation{
+						MaxLength: maxLength,
+					},
+				}},
+			},
+			ValueValidation: &schema.ValueValidation{
+				MaxProperties: maxLength,
+			},
+			Extensions: schema.Extensions{
+				XValidations: apiextensions.ValidationRules{
+					{
+						Rule: rule,
+					},
+				},
+			},
+		}
+	}
+}
+
+func genAllMaxNestedSpecWithRootRule(rule string) func(maxLength *int64) *schema.Structural {
+	return func(maxLength *int64) *schema.Structural {
+		return &schema.Structural{
+			Generic: schema.Generic{
+				Type: "array",
+			},
+			Items: &schema.Structural{
+				Generic: schema.Generic{
+					Type: "object",
+					AdditionalProperties: &schema.StructuralOrBool{Structural: &schema.Structural{
+						Generic: schema.Generic{
+							Type: "object",
+						},
+						ValueValidation: &schema.ValueValidation{
+							Required:      []string{"required"},
+							MaxProperties: maxLength,
+						},
+						Properties: map[string]schema.Structural{
+							"required": {
+								Generic: schema.Generic{
+									Type: "string",
+								},
+							},
+							"optional": {
+								Generic: schema.Generic{
+									Type: "string",
+								},
+							},
+						},
+					}},
+				},
+				ValueValidation: &schema.ValueValidation{
+					MaxItems:      maxLength,
+					MaxProperties: maxLength,
```

Comment:
Why both `MaxItems` and `MaxProperties` on the same schema? Since this is an "object", the `MaxProperties` should be the only one doing anything.

Reply:
Oh, you need to move the `MaxItems` up to the surrounding array again here too.